packages/input_schema/src/types.ts

```diff
@@ -0,0 +1,116 @@
+type CommonFieldDefinition<T> = {
+    title: string;
+    description: string;
+    default?: T;
+    prefill?: T extends boolean ? never : T;
+    example?: T;
+    nullable?: boolean;
+    sectionCaption?: string;
+    sectionDescription?: string;
+}
+
+export type StringFieldDefinition = CommonFieldDefinition<string> & { type: 'string' } & {
+    // json is specified in tests, but not in docs
+    editor: 'textfield' | 'textarea' | 'javascript' | 'python' | 'select' | 'datepicker' | 'hidden' | 'json';
+    pattern?: string;
+    minLength?: number;
+    maxLength?: number;
+    enum?: string[]; // required if editor is 'select'
+    enumTitles?: string[]
+    isSecret?: boolean;
+}
+
+export type BooleanFieldDefinition = CommonFieldDefinition<boolean> & { type: 'boolean' } & {
+    editor?: 'checkbox' | 'hidden';
+    groupCaption?: string;
+    groupDescription?: string;
+}
+
+export type IntegerFieldDefinition = CommonFieldDefinition<number> & { type: 'integer' } & {
+    editor?: 'number' | 'hidden';
+    maximum?: number;
+    minimum?: number;
+    unit?: string;
+}
+
+export type ObjectFieldDefinition = CommonFieldDefinition<object> & { type: 'object' } & {
+    editor: 'json' | 'proxy' | 'hidden';
+    patternKey?: string;
+    patternValue?: string;
+    maxProperties?: number;
+    minProperties?: number;
+}
+
+export type ArrayFieldDefinition = CommonFieldDefinition<Array<unknown>> & { type: 'array' } & {
+    editor: 'json' | 'requestListSources' | 'pseudoUrls' | 'globs' | 'keyValue' | 'stringList' | 'select' | 'hidden';
+    placeholderKey?: string;
+    placeholderValue?: string;
+    patternKey?: string;
+    patternValue?: string;
+    maxItems?: number;
+    minItems?: number;
+    uniqueItems?: boolean;
+    items?: unknown;
+}
+
+type AllTypes = StringFieldDefinition['type']
+    | BooleanFieldDefinition['type']
+    | IntegerFieldDefinition['type']
+    | ObjectFieldDefinition['type']
+    | ArrayFieldDefinition['type']
+
+export type MixedFieldDefinition = CommonFieldDefinition<never> & {
+    type: AllTypes[] | readonly AllTypes[];
+    editor: 'json'
+}
+
+export type FieldDefinition = StringFieldDefinition
+    | BooleanFieldDefinition
+    | IntegerFieldDefinition
+    | ObjectFieldDefinition
+    | ArrayFieldDefinition
+    | MixedFieldDefinition
+
+// should not be present, this type is for invalid schema
+type NeverFieldDefinition = CommonFieldDefinition<never> & { type?: undefined, editor?: string }
+
+type FieldDefinitionToUnchecked<T extends FieldDefinition | NeverFieldDefinition> = Partial<
+    Omit<T, 'type'> & { type: string | string[] | readonly string[] }
```

Comment:
I kind of hoped that we could keep discriminating union there - but... 😞 
it would break all the existing things 😅 Mainly because we are not freezing the config objects (`as const`) so they are inferred as `string`, not constants (`'string', 'number', 'array'`, etc.)